Persona back-end integration

spratt opened the pull request:
We now perform all the necessary steps on the server-side to fully authenticate users with the Persona protocol.

Once a user is authenticated, their email is stored in the session variables.  Users now comment either as "Anonymous" (if not signed in) or as their email (if signed in).

closes #39
closes #104

psimonyi committed:
Put the Persona login button on the add-code page

	new file:   src/client/css/persona-buttons.css
	modified:   src/client/index.html
	modified:   src/client/index.js
	new file:   src/client/login.js
	modified:   src/client/view.css

spratt committed:
Persona works again

spratt committed:
Fixed persona style

spratt committed:
Fixed coding standard js.1

spratt committed:
Now sending assertion to server

spratt committed:
Now using sessions

spratt committed:
Now validating persona assertions

spratt committed:
Now passing in host, clientPort to requestHandlers

for persona assertion validation

spratt committed:
Now only parsing cookies on post requests

spratt committed:
Now commenting as anonymous or email

spratt committed:
Now indicating to user who they are commenting as

spratt committed:
Comments now appear in chronological order

spratt committed:
Now whitelisting code objects before storing

spratt committed:
Added a little documentation, removed debug code

spratt committed:
Now indicating if user is signed in

spratt committed:
Comment box now lines up with code

spratt commented:
Every weekend I ask you guys to pull another fuck-you-sized request.  I am sorry about this.

cdelahousse commented on src/client/index.html:
```diff
@@ -4,17 +4,31 @@
-    <h1><a href="index.html">CodeReview</a></h1>
+    <h1 id="title"><a href="index.html">CodeReview</a></h1>
+      Try our
+      <a href="javascript:(function(){window.codeReview&&window.codeReview.exit();var a=document.createElement('script');a.src='//review.steamedpears.com/bookmarklet/bookmarklet.js';(document.head||document.body).appendChild(a)})();">bookmarklet</a>
```
Good call, moving the bookmarklet up.

spratt replied:
Thanks!

psimonyi replied:
This URL exists elsewhere, right? We should somehow pull it from wherever it lives, or else remove it from the old place and make this canonical. IOW, this thing should exist in one place in the repository. [non-blocking, file-a-bug]

cdelahousse commented on src/client/scripts/login.js:
```diff
@@ -0,0 +1,79 @@
+    label_span = $('#user-label');
+  });
+
+  $.getScript('https://login.persona.org/include.js', ready);
```
Couldn't we just pull this in using RequireJS instead of with jQuery? 
http://requirejs.org/docs/api.html#config-paths

spratt replied:
It was part of @psimonyi's original implementation.  I think I'd prefer to use RequireJS for consistency, so I'll see if I can get it working that way instead.

cdelahousse commented on src/client/scripts/login.js:
```diff
@@ -0,0 +1,79 @@
+  $('#persona-logout').click(function login_persona_click() {
+    navigator.id.logout();
+  });
+  function ready() {
```
Is this a good idea to separate the `ready` function from where it's set as a callback on line 21. Why not just define this inline?

spratt replied:
Part of the original implementation.  This was probably more similar to our conventions at the time.

cdelahousse commented on src/client/scripts/view.js:
```diff
@@ -123,6 +123,7 @@ define([
     $('#line-end-num').text(end+1);
     editor.setDiffSelected(start,end);
     var pos = editor.getLinePosition(start);
+    console.log(pos);
```
Is this debug code? You may want to strip this out.

spratt replied:
Good find.

cdelahousse commented on src/server/dev_mode.js:
```diff
@@ -12,15 +12,14 @@ var proxy = require('proxy-middleware');
   var proxyRoutes = {
-    '/do' : 'http://localhost:' + serverPort,
-    '/'   : 'http://localhost:' + staticPort
+    '/do' : 'http://' + host + ':' + serverPort,
```
Nice.

spratt replied:
I like this better...but it will only ever be called with host='localhost', since this is just for dev mode.  Oh well.

cdelahousse commented on src/server/index.js:
```diff
@@ -59,7 +65,9 @@ for (var route in getRoutes) {
-app.use(connect.bodyParser());
+app.use(connect.bodyParser())
+  .use(connect.cookieParser())
+  .use(connect.session({secret: 'keyboard kitties'}));
```
Giggles.

spratt replied:
There's probably a good convention for what we should use as our secret key.  This is basically connect.js's default.

cdelahousse commented on src/server/requestHandlers.js:
```diff
@@ -121,14 +122,19 @@ exports.commentCount = function commentCount(request, response) {
 * Setters                                                                     *
 ******************************************************************************/
-exports.newcode = function newcode(request, response) {
+function newcode(request, response) {
```
I'm not a fan of removing the exports assignments form the function declaration. If you keep it, I know that as I read the code, I can expect this to be a public facing API instead of learning about this near the bottom of the file.

spratt replied:
The reason I did this was to pass in the host and clientPort arguments, and I didn't want to put all of the functions inside of the function I used for that.  See line 199.

cdelahousse replied:
Ah.

spratt replied:
I'm not super happy with it either.  Suggestions?

spratt replied:
I have a better idea.  Commit incoming.

spratt committed:
Cleaned up requestHandlers exports

spratt committed:
Realigned all the elements again

spratt committed:
Removed debug code

spratt committed:
now creating ready function within the getScript call

spratt committed:
Now using requireJS in src/client/login.js

spratt committed:
Pretty buttons

spratt committed:
General visual cleanup

spratt committed:
Now using blue persona login button

spratt commented:
[CRITICAL] I've found a major bug that causes the db to lose the lang attribute on code objects.  Do not merge this until fixed.

spratt committed:
Fixed critical bug

No longer submitting code form when opening comments
No longer accepting code without lang

spratt commented:
Fixed.  The problem was a cascade of bugs.

First, if you were visiting a piece of code you hadn't just submitted, the language select box would be unpopulated.  If you expanded comments on a line, it would submit the code form again.  Our `/do/newcode` request wasn't checking for a valid lang, so it would succeed and create a new code object.  This would change the url of the page to the new ID.  Now, if you refreshed the page you would get the new code with unspecified lang, which made it look like the db was dropping the lang attribute.

Fixes:
- `/do/newcode` now checks if the lang attribute is properly set.
- Expanding the comments on a line doesn't submit the code form.

spratt commented on src/server/requestHandlers.js:
```diff
@@ -156,17 +170,89 @@ exports.newcomment = function newcomment(request, response) {
+******************************************************************************/
+
+module.exports = function(host, clientPort) {
+  public_api.login = function login(request, response) {
```
This is the most complex part of the back-end now.  I'm not sure how I feel about this.

psimonyi replied:
Perhaps this belongs in its own module. [non-blocking]

psimonyi replied:
Made a bug.

spratt committed:
Merge branch 'master' into personaServer

spratt committed:
Fixed diff line number problem

cdelahousse commented:
It looks good and you've fixed the bug. Can I merge this bitch now?

spratt commented:
May as well let @psimonyi have a look.

spratt committed:
Updated HTTP API documentation

spratt committed:
Updated http_api.md to include HTTP responses

spratt committed:
Updated API doc to include invalid referer on newcomment

psimonyi commented on src/client/index.html:
```diff
@@ -4,17 +4,31 @@
+        Tutorial
+      </button>
+    </aside>
+    <br clear="all">
```
:(

spratt replied:
What's wrong with this?

psimonyi replied:
On Wed, 2013-04-03 at 16:43 -0700, Simon David Pratt wrote:

> What's wrong with this?
> It's very much about the presentation, not the content.

psimonyi replied:
[non-blocking]

spratt replied:
Most of the HTML is about presentation?  It's basically a structure into which we inject our content.

psimonyi commented on src/server/requestHandlers.js:
```diff
@@ -1,9 +1,15 @@
 /******************************************************************************
+* The methods on this object will be exported as the public api.              *
+******************************************************************************/
+var public_api = {};
```
Does one line of code really need a 3-line comment heading? I thought the starry stuff was for sections of code, but this comment isn't about a section.

cdelahousse replied:
Agreed
On Apr 3, 2013 6:40 PM, "Peter Simonyi" notifications@github.com wrote:

> In src/server/requestHandlers.js:
> 
> >  var url = require('url');
> >  var uuid = require('node-uuid');
> >  var db = require('redis').createClient();
> > 
> >  /******************************************************************************
> > +\* The methods on this object will be exported as the public api.              *
> > +******************************************************************************/
> > +var public_api = {};
> 
> Does one line of code really need a 3-line comment heading? I thought the
> starry stuff was for sections of code, but this comment isn't about a
> section.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/CarletonU-COMP2406-W2013/Steamed-Pears/pull/111/files#r3646631
> .

psimonyi commented on src/server/requestHandlers.js:
```diff
@@ -156,17 +173,88 @@ exports.newcomment = function newcomment(request, response) {
+    }
+    var assertion = request.body.assertion;
+    var content = 'assertion=' + assertion + '&audience=http://';
+    content += host + ':' + clientPort;
```
In Python, I'd have `'assertion={asrt}&audience=http://{host}:{port}'.format(asrt=assertion, host=host, port=port)`. Isn't there something like this in JS?

psimonyi commented on src/server/requestHandlers.js:
```diff
@@ -156,17 +173,88 @@ exports.newcomment = function newcomment(request, response) {
+      return error(response, 400, 'Invalid assertion');
+    }
+    var assertion = request.body.assertion;
+    var content = 'assertion=' + assertion + '&audience=http://';
```
What if someone provides an assertion like `1&audience=http://mysite.com`?

psimonyi replied:
We should probably URL-encode the assertion.

psimonyi commented on src/server/requestHandlers.js:
```diff
@@ -156,17 +173,88 @@ exports.newcomment = function newcomment(request, response) {
+                'Content-Type': 'application/x-www-form-urlencoded'}
+    },function(auth_response) {
+      var data_str = '';
+      // assume uft8
```
Can't we use Connect's body_parser thing here too? I thought it handled the encoding and waiting for all data things.

spratt replied:
body_parser is meant to go into the connect call chain and parse the body of an incoming request, not the body of an incoming response to a request we sent.  It might be able to do it, but it's not what it's intended for.

psimonyi commented on http_api.md:
```diff
@@ -0,0 +1,187 @@
+Required parameters:
+
+* `text`: The (code) text of the code object.
+* `lang': The programming language in which the text is written.
```
Closing quote: Markdown, not LaTeX.

spratt committed:
Whittled down excessively large comment

psimonyi commented on src/client/scripts/login.js:
```diff
@@ -0,0 +1,81 @@
+
+  require([
+    'https://login.persona.org/include.js'
+  ], function persona_ready() {
```
Why not just include this in the dependencies for `define` and save on a callback and indentation level?

psimonyi commented on src/client/scripts/login.js:
```diff
@@ -0,0 +1,81 @@
+    if(user_display !== null) {
+      user_display.text('Anonymous');
+    }
+    if(button !== null) {
```
Could we have spaces after the `if`s please?

psimonyi commented on src/server/requestHandlers.js:
```diff
@@ -121,14 +129,22 @@ exports.commentCount = function commentCount(request, response) {
-  var obj = request.body;
-  if (obj === null || !isValidString(obj.text)) {
+  var fields = request.body;
+  if (fields === null || !isValidString(fields.text)) {
```
Not just `if (!fields || ...`? As it is, the empty string is valid code. And what if request.body is undefined?

spratt committed:
Fixed errant apostrophe

psimonyi commented on src/server/requestHandlers.js:
```diff
@@ -94,9 +99,12 @@ exports.commentsOnLine = function commentsOnLine(request, response) {
   var query = url.parse(request.url, true).query;
   var code_id = query.code_id;
+  if(code_id === undefined) {
+    return error(response, 400, 'Invalid code id');
```
Shouldn't this return invalid code id when the code id isn't in the database too? It seems there should be a difference between a code id that doesn't exist, and a code id that exists but has no comments.

psimonyi replied:
This got changed somewhat in another commit, and this behaviour is consistent within CR, so it's fine.

psimonyi commented on src/server/requestHandlers.js:
```diff
@@ -156,17 +171,88 @@ exports.newcomment = function newcomment(request, response) {
+      auth_response.on('end',function() {
+        var data_ob = JSON.parse(data_str);
+        if (auth_response.statusCode === 200 &&
+            data_ob !== null &&
```
Apparently `JSON.parse` will throw an exception for invalid JSON. It would only return `null` if the JSON text is acutally `null`. [Reference](https://developer.mozilla.org/en-US/docs/JavaScript/Reference/Global_Objects/JSON/parse)

spratt committed:
Now requiring persona include.js in dependencies

spratt committed:
Coding standard js.12

spratt committed:
Now using string format and url encode

spratt committed:
Now catching JSON.parse exceptions

spratt committed:
Merged

psimonyi commented on src/server/requestHandlers.js:
```diff
@@ -156,17 +181,94 @@ exports.newcomment = function newcomment(request, response) {
+module.exports = function(host, clientPort) {
+  public_api.login = function login(request, response) {
+    if(request === null ||
+       request.body === null ||
```
With Connect, it appears request.body will never be null. If it's any falsy value, body_parser sets it to an empty object. I think `if (!request.body.assertion)` will do.

spratt committed:
Now checking truthy/falsy values

psimonyi commented on src/client/index.html:
```diff
@@ -4,17 +4,31 @@
+        <span id="user-email"></span>
+      </span>
+      <button id="persona-login" class="persona-button">
+        <span id="persona-button-text">Sign in</span>
```
Why do you need this extra span?

psimonyi replied:
So it's my code. We still don't need it.

spratt replied:
CSS.

psimonyi replied:
Oops. Right.

psimonyi commented on src/server/requestHandlers.js:
```diff
@@ -130,37 +156,110 @@ exports.newcode = function newcode(request, response) {
+          return error(response, 500, 'Invalid verification server response');
+        }
+        if (auth_response.statusCode === 200 &&
+            data_ob && data_ob.status &&
```
There's no need to check that both `data_ob.status` is not falsy and that it is `'okay'`.

spratt committed:
Checking truthy by checking value of persona status

psimonyi commented:
`/do/login` returns a 400 Bad Request. The log file shows `===ERROR=== 400 Invalid assertion`, and also a whole bunch of errors connecting to redis.

psimonyi commented:
The database server doesn't seem to be started by the scripts in this branch. Is this affecting `/do/login`?

spratt commented:
bin/start_dev.sh seems to start the database for me.  But I am also getting a 400 on /do/login.  Investigating.

spratt commented:
I have inverted logic in the sanity check for /do/login.  Fixing now.

spratt committed:
Fixed inverted logic in /do/login

spratt committed:
Merged with language cleanup
